### PR TITLE
Adding updated noisy channel detection and trend removal in EEG data

### PR DESCRIPTION
**Pull request by AamnaLawrence**
The updated noisy channels code detects noisy channels by testing them on the basis of deviation, correlation, random sampling consensus, low signal to noise ratio, high-frequency noise, flatness and Nan values. This function is iteratively called when performing robust referencing.
The remove trend function filters the data using high pass and high pass sinc filters. It can also perform local detrending of data using. This function is used in PREP pipeline and iteratively called when performing robust referencing. 
The tests for both the above-mentioned functionalities are also provided. 
@adam2392

### CODE REVIEW

File: pyprep/test_find_noisy_channels.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+for iter in range(0, iterations):
+    raw.info["bads"] = bads
+    raw.interpolate_bads()
+    nd = NoisyChannels(raw)
+    nd.find_all_noisy_channels(ransac=True)
+    bads = nd.get_bads()
+
+# make sure no bad channels exist in the data
+if bads != []:
+    raw.drop_channels(ch_names=bads)
```
Comment (adam2392) [inline pyprep/test_find_noisy_channels.py]: These should all be in a test function.

File: pyprep/test_removeTrend.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    signal_trend = 2 + 1.5 * np.linspace(0, 1, npoints) + signal
+    signal_detrend = removeTrend.removeTrend(
+        signal_trend, detrendType="Local detrend", sample_rate=100
+    )
+    error3 = signal_detrend - signal
+    assert np.sqrt(np.mean(error3 ** 2)) < 0.1
+
+
+test_highpass()
+test_detrend()
```
Comment (adam2392) [inline pyprep/test_removeTrend.py]: Same here. These are just called when you run: "pytest ./test/"

File: pyprep/test_find_noisy_channels.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    nd.find_bad_by_ransac()
+    bads = nd.bad_by_ransac
+    assert bads == raw_tmp.ch_names[0:6]
+
+
+test_find_bad_by_nan_flat(raw)
+test_find_bad_by_ransac(raw)
+test_find_bad_by_hf_noise(raw)
+test_find_bad_by_correlation(raw)
+test_find_bad_by_SNR(raw)
```
Comment (adam2392) [inline pyprep/test_find_noisy_channels.py]: Unnecessary.

**Review by adam2392 — CHANGES_REQUESTED** [review]
Stylistic Changes:

1. Setup of tests: use the "conftest.py" file and pytest fixtures framework. Any setup of data structures can be done in the conftest.py file, and then imported over using pytest fixtures. Google this and see what I mean. We can discuss in person. 
2. Add a "scikit-learn" fit() style function if find_noisy_channels is going to be called and do an entire pipeline with default parameters. If separate functions need to be called, then not needed.

File: pyprep/find_noisy_channels.py [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+
+    def __init__(self, raw):
+
+        """Initialize the class."""
+
+        # Make sure that we got an MNE object
+        assert isinstance(raw, mne.io.BaseRaw)
+
+        self.raw_mne = raw.copy()
+        self.EEGData = self.raw_mne.get_data(picks="eeg")
```
Comment (Nick3151) [inline pyprep/find_noisy_channels.py:33]: Extract eeg data here:
self.raw_mne.pick_types(eeg=True, stim=False)

File: pyprep/find_noisy_channels.py [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+                )
+            )
+            print(
+                "\n{} by ransac: {}".format(len(self.bad_by_ransac), self.bad_by_ransac)
+            )
+
+        return bads
+
+    def find_all_noisy_channels(self, ransac=True):
+
```
Comment (Nick3151) [inline pyprep/find_noisy_channels.py:110]: Please change this function name to find_all_bads so I can easily compare your code with Noisydata

**Review by Nick3151 — CHANGES_REQUESTED** [review]
Please add the extract EEG data line as mentioned, otherwise it will cause a bug with channel index
Please change the function name as mentioned so I can easily switch between your version and Noisydata

**Pull request closed without merging**